TheraLang.BLL/Services/ProjectService.cs

```diff
@@ -0,0 +1,192 @@
+                    .ForMember(p => p.DonationTargetSum, opt => opt.MapFrom(p => p.DonationTarget)))
+                .CreateMapper();
+            var projectsDto = mapper.Map<IEnumerable<Project>, IEnumerable<ProjectDto>>(projects);
+
+            return projectsDto;
+        }
+
+        public IEnumerable<ProjectDto> GetAllNewProjects()
+        {
+            var projects = _uow.Repository<Project>().Get().Where(i => i.StatusId == 0);
+
+            var mapper = new MapperConfiguration(cfg => cfg.CreateMap<Project, ProjectDto>()
+                    .ForMember(p => p.DonationTargetSum, opt => opt.MapFrom(p => p.DonationTarget)))
+                .CreateMapper();
+            var projectsDto = mapper.Map<IEnumerable<Project>, IEnumerable<ProjectDto>>(projects);
+
+            return projectsDto;
+        }
+
+        public IEnumerable<ProjectDto> GetProjectsByStatus(int status)
+        {
+            var projects = _uow.Repository<Project>().Get().Where(i => i.StatusId == (ProjectStatus)status).ToArray();
+
+            var mapper = new MapperConfiguration(cfg => cfg.CreateMap<Project, ProjectDto>()
+                    .ForMember(p => p.DonationTargetSum, opt => opt.MapFrom(p => p.DonationTarget)))
+                .CreateMapper();
+            var projectsDto = mapper.Map<IEnumerable<Project>, IEnumerable<ProjectDto>>(projects);
+
+            return projectsDto;
+        }
+
+        public async Task Add(ProjectDto projectDto, Guid userId)
+        {
+            var mapper = new MapperConfiguration(cfg => cfg.CreateMap<ProjectDto, Project>()
+                .ForMember(p => p.DonationTarget, opt => opt.MapFrom(src => src.DonationTargetSum))
+                .ForMember(p => p.IsActive, opt => opt.MapFrom(src => true)))
+                .CreateMapper();
+
+            var project = mapper.Map<ProjectDto, Project>(projectDto);
+
+            var newParticipant = new ProjectParticipation
+            {
+                Role = MemberRole.ProjectOwner,
+                CreatedById = userId,
+                Status = ProjectParticipationStatus.Approved,
+                Project = project,
+            };
+            try
+            {
+                await _uow.Repository<Project>().Add(project);
+                await _uow.SaveChangesAsync();
+                await _uow.Repository<ProjectParticipation>().Add(newParticipant);
+                await _uow.SaveChangesAsync();
+            }
+            catch (Exception e)
+            {
+                e.Data["project"] = projectDto;
+                throw;
+            }
+        }
+
+        public async Task Delete(int id)
+        {
+            try
+            {
+                var project = _uow.Repository<Project>().Get().SingleOrDefault(p => p.Id == id);
+                if (project == null)
+                {
+                    throw new NullReferenceException($"Error while deleting project. Project with id {nameof(id)}={id} not found");
```

Comment:
this is for sure not NullReferenceException, i guess we should create or custom exception, or use DataException

Reply:
Why not NullReferenceException?
SingleOrDefault can return null

Reply:
> this is for sure not NullReferenceException, i guess we should create or custom exception, or use DataException

Why not NullReferenceException?
SingleOrDefault can return null

Reply:
Agree, SingleOrDefault can return null, tho we will get NullReferenceException only if we call some props form object.
but right now we just say if project is null then return NullReferenceException, it's confusing and not correct! for example, u will see this error in logs, and try to find fix for it, to do so you will try to search some calls to props of an object that is null, and will not expect this.
About the custom exception, since we were trying to get element, and our db doesn't have it, it's either NotFound or some worng input parameter, also it can be treated as Validation error, and for validation errors it's preferable to use Custom exception.